fsql/deser.py

```diff
@@ -75,28 +104,37 @@ def detect_format(self, url: str) -> InputFormat:
             return InputFormat.from_url(url.split(".")[-1])
 
     @abstractmethod
-    def read_single(self, partition: Partition, fs: AbstractFileSystem) -> Any:  # TODO generics on the Any
+    def read_single(self, partition: Partition, fs: AbstractFileSystem) -> PartitionReadOutcome:
         raise NotImplementedError("abc")
 
     @abstractmethod
-    def concat(self, data: Iterable[Any]) -> Any:  # TODO generics on the Any
+    def concat(self, outcomes: Iterable[DataObject]) -> DataObject:
         raise NotImplementedError("abc")
 
     def read_and_concat(
         self, partitions: Iterable[Partition], fs: AbstractFileSystem
-    ) -> Any:  # TODO generics on the Any
-        with ThreadPoolExecutor(max_workers=32) as tpe:  # TODO configurable
-            data = tpe.map(partial(self.read_single, fs=fs), partitions)
-        return self.concat(data)
-
-
-class PandasReader(DataReader):
+    ) -> Union[DataObject, Tuple[DataObject, Iterable[PartitionReadFailure]]]:
```

Comment:
hmm, I really do not like this 😃 Could you elaborate a bit on the possible solution described in the comments below? 
@ampx-mg Do you have any opinion?

Reply:
ideally, instead of boolean parameter lazy_errors, you would have a class ErrorHandler with two typevars [DataObject, ReaderOutcome], where ReaderOutcome would be bound by DataObject and Tuple[DataObject, Iterable[PartitionReadFailure]], and the ReaderOutcome typevar would be also the return type of the function. The class would have just one method, `handle(Tuple[DataObject, Iterable[...]]): ReaderOutcome`, and there would be two instances/implementations of the class, each collapsing the output type to one of the two options and having the `handle` implemented accordingly (either identity or `reader_outcome[0]`). Furthermore, you would use the `error_handler` instance inside the `read_single` to match to decide whether to raise or not (possibly, the class itself would be responsible for `Exception -> PartitionReadFailure` transformation, and it _itself_ would do the raise in the non-lazy case).

the problem was that I was not able to make it work, that is all...

I understand you don't like the Union, neither do I, in particular because its sorta breaking change of api... however, the only other options I came up with was creating a wholly new method which seems an unjustified api bloat to me, and always returning the Tuple which is even more breaking

Reply:
I do agree however as Vojta has described the issue I dont see a better way out unfortunately

Reply:
I have an issue with the fact that the proposal returns either single element (data) or two (failures added). That's why I would rather return tuple in all cases, but it's not wise exactly from the reason you mentioned, this would be really major breaking change. Therefore I would be probably happier with a solution that would omit failures completely (either raise exception or log errors/warnings). Although I understand the possibility to analyze read failures would not be possible.

Regarding the solution above, this would also be kind of breaking change, right? Or how would you access read failure in such case?

Reply:
the "ideal" solution I described would not be breaking -- the function with default arguments would be returning DataObject, but with error_handling=LazyErrorsHandler would return the Tuple. So in general the function would still be returning "Union", but the mypy would be able to tell which one as long as the type of the `error_handling` parameter is known to it. That would be, to me, the most elegant

I really don't like the exception throwing -- I could imagine doing it in the way that we keep the return type DataObject, and only change whether we raise on the first exception, or whether we collect all exceptions and only raise with the final one. But that loses the ability to propagate the DataObject itself in the lazy case -- which is the whole point of this exercise :) Logging is surely bad, you need to be able to programmatically assess how many files etc you've lost...

Reply:
ok, thanks for the explanation. I still do not like it much 😄 (as you wrote, it would still be kind of Union, although more elegant), but can live with it.

Regarding logging - it depends. When you are aware of it and know that for example passing parameter `columns=[...]` will read only those files that contain all those columns, I do not think it's necessary (but agree it might be useful sometimes). Then, instead of true/false for lazy errors, you could even have e.g. options like `"raise", "warn", "skip"` to have also option to skip the errors completely. 

The last idea (just asking for your view in general, it would break the api similarly as tuple would ofc) - what do you think about a dedicated object that would have a data field + some optional fields for errors for example? I know that would be a breaking change which is sth we do not want. But despite this, would it make sense or not in your opinion? (The advantage could be that you could always access data in the same way like `returned_obj = ...; returned_obj.data`, and it would give you an option to look at `returned_obj.failures` if needed).

Anyway, if you both are ok with the current solution, let's merge 🙂

Reply:
yes, it would be an improvement imo 🙂 but if it would take too much effort due to some mypy issues and so on we can go with the current solution.

Reply:
ok so I did introduce a dedicated object, DataObjectRich, and I liked the name and the resulting signature so much (DataObjectExt would also work) that I let it there despite mypy complaining as I feared. I managed to get rid of one of the complaints by ugrading mypy, but alas, some type: ignore do remain :( 

that concludes all planned changes, @sondracek or @sxooler, feel free to review the latest commit and possibly merge